src/tritonserver.cc

```diff
@@ -2157,6 +2157,112 @@ TRITONSERVER_ServerStop(TRITONSERVER_Server* server)
+        mode = "POLL";
+        break;
+      default:
+        mode = "UNKNOWN";
+        break;
+    }
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_UNSUPPORTED,
+        (std::string("Register API is unsupported in ") + mode +
+         " model control mode")
+            .c_str());
+  }
+  if ((name_mapping == nullptr) && (mapping_count != 0)) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INVALID_ARG,
+        "model mappings are not provided while mapping count is non-zero");
+  }
+  bool is_directory = false;
+  auto status = triton::core::IsDirectory(repository_path, &is_directory);
+  if (!status.IsOk() || !is_directory) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INVALID_ARG,
+        (std::string("failed to register '") + repository_path +
+         "', repository not found")
+            .c_str());
+  }
+
+  RETURN_IF_STATUS_ERROR(lserver->AddModelRepositoryPath(repository_path));
+  std::unordered_map<std::string, std::string> model_mapping;
+
+  for (size_t i = 0; i < mapping_count; ++i) {
+    auto mapping =
+        reinterpret_cast<const tc::InferenceParameter*>(name_mapping[i]);
+    auto subdir = mapping->Name();
+
+    if (mapping->Type() != TRITONSERVER_PARAMETER_STRING) {
+      return TRITONSERVER_ErrorNew(
+          TRITONSERVER_ERROR_INVALID_ARG,
+          std::string(
+              "Mapped model name must be a string, found "
+              "another type for " +
+              subdir)
+              .c_str());
+    }
+
+    auto model_name =
+        std::string(reinterpret_cast<const char*>(mapping->ValuePointer()));
+
+    tc::Status status =
+        lserver->AddModelMapping(model_name, repository_path, subdir);
+    if (!status.IsOk()) {
+      TRITONSERVER_ServerUnregisterModelRepository(server, repository_path);
+      return TRITONSERVER_ErrorNew(
+          TRITONSERVER_ERROR_INVALID_ARG,
+          (std::string("failed to register '") + repository_path +
+           "', there is conflicting mapping for '" + std::string(model_name) +
+           "'")
+              .c_str());
+    }
+  }
+
+  return nullptr;  // Success
+}
+
+TRITONSERVER_DECLSPEC TRITONSERVER_Error*
+TRITONSERVER_ServerUnregisterModelRepository(
+    TRITONSERVER_Server* server, const char* repository_path)
+{
+  tc::InferenceServer* lserver = reinterpret_cast<tc::InferenceServer*>(server);
+  if (lserver->GetModelControlMode() != tc::ModelControlMode::MODE_EXPLICIT) {
+    std::string mode;
+    switch (lserver->GetModelControlMode()) {
+      case tc::ModelControlMode::MODE_NONE:
+        mode = "NONE";
+        break;
+      case tc::ModelControlMode::MODE_POLL:
+        mode = "POLL";
+        break;
+      default:
+        mode = "UNKNOWN";
+        break;
+    }
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_UNSUPPORTED,
+        (std::string("Unregister API is unsupported in ") + mode +
+         " model control mode")
+            .c_str());
+  }
```

Comment:
The control mode check should be moved to `ModelRepositoryManger`, similar to the check in `ModelRepositoryManager::LoadUnloadModel`, we can update the test to expect error message like `repository register is not allowed if model control mode is EXPLICIT` for simplicity.

Reply:
Done.